lightning/src/ln/channelmanager.rs

```diff
@@ -5533,33 +5575,72 @@ where
 							} else { None };
 
-							Some(MonitorUpdateCompletionAction::EmitEventAndFreeOtherChannel {
-								event: events::Event::PaymentForwarded {
-									fee_earned_msat,
-									claim_from_onchain_tx: from_onchain,
-									prev_channel_id: Some(prev_outpoint.to_channel_id()),
-									next_channel_id: Some(next_channel_outpoint.to_channel_id()),
-									outbound_amount_forwarded_msat: forwarded_htlc_value_msat,
-								},
-								downstream_counterparty_and_funding_outpoint:
-									if let Some(node_id) = next_channel_counterparty_node_id {
-										Some((node_id, next_channel_outpoint, completed_blocker))
-									} else {
-										// We can only get `None` here if we are processing a
-										// `ChannelMonitor`-originated event, in which case we
-										// don't care about ensuring we wake the downstream
-										// channel's monitor updating - the channel is already
-										// closed.
-										None
+							let chan_to_release =
+								if let Some(node_id) = next_channel_counterparty_node_id {
+									Some((node_id, next_channel_outpoint, completed_blocker))
+								} else {
+									// We can only get `None` here if we are processing a
+									// `ChannelMonitor`-originated event, in which case we
+									// don't care about ensuring we wake the downstream
+									// channel's monitor updating - the channel is already
+									// closed.
+									None
+								};
+
+							if definitely_duplicate && startup_replay {
+								// On startup we may get get redundant claims which are related to
+								// monitor updates still in flight. In that case, we shouldn't
+								// immediately free, but instead let that monitor update complete
+								// in the background.
+								#[cfg(debug_assertions)] {
+									let background_events = self.pending_background_events.lock().unwrap();
+									// There should be a `BackgroundEvent` pending...
+									assert!(background_events.iter().any(|ev| {
+										match ev {
+											// to apply a monitor update that blocked channel,
+											BackgroundEvent::MonitorUpdateRegeneratedOnStartup {
+												funding_txo, ..
+											} => *funding_txo == claiming_chan_funding_outpoint,
+											// or the channel we'd unblock is already closed,
+											BackgroundEvent::ClosedMonitorUpdateRegeneratedOnStartup((funding_txo, ..))
+												=> *funding_txo == next_channel_outpoint,
+											// or the monitor update has completed and will unblock
+											// immediately once we get going.
+											BackgroundEvent::MonitorUpdatesComplete {
+												channel_id, ..
+											} =>
+												*channel_id == claiming_chan_funding_outpoint.to_channel_id(),
+										}
+									}), "{:?}", *background_events);
+								}
+								None
+							} else if definitely_duplicate {
+								if let Some(other_chan) = chan_to_release {
+									Some(MonitorUpdateCompletionAction::FreeOtherChannelImmediately {
```

Comment:
I know that we're certain that the `Channel` has seen this preimage at this stage, but I'm having trouble convincing myself that the upstream monitor update has also fully completed. It looks like we'll return `::DuplicateClaim` from the upstream `Channel` if the htlc claim is already in the holding cell, for example -- I think that means the monitor update is not released? Any context I'm missing here?